pkg/nodeidentity/openstack/identify.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/kops/pkg/nodeidentity"
+)
+
+// MetadataLatest url for latest metadata which contains instance metadata
+const MetadataLatest string = "http://169.254.169.254/openstack/latest/meta_data.json"
+
+// nodeIdentifier identifies a node
+type nodeIdentifier struct {
+}
+
+// New creates and returns a nodeidentity.Identifier for Nodes running on OpenStack
+func New() (nodeidentity.Identifier, error) {
+	return &nodeIdentifier{}, nil
+}
+
+// IdentifyNode queries OpenStack for the node identity information
+func (i *nodeIdentifier) IdentifyNode(ctx context.Context, node *corev1.Node) (*nodeidentity.Info, error) {
+	metadata, err := getLocalMetadata()
```

Comment:
Ah - sorry - there's a gotcha here.  We need to get the instancegroup of `node`, which usually won't be our node.  The idea is that we can't trust the nodes to self-report (or we don't want to).

Looking at some of the existing openstack code, I think you need to do a ListInstances request and then get the `KopsInstanceGroup` metadata (nice forward thinking of adding that metadata already).  You can identify the instance using the node.Spec.ProviderID which should (hopefully) be set on the node by openstack cloudprovider.

I'm not sure if the instance metadata is trivially writeable from the instance, for example by posting to http://169.254.169.254.  If it is we might need to beef up the security somehow, but I think the ProviderID is currently unprotected anyway...

Reply:
okay then its bigger change :) I thought this is going to be easy!